src/app/src/main/kotlin/org/imperial/mrc/hint/exceptions/ExceptionHandler.kt

```diff
@@ -14,48 +13,70 @@ import org.springframework.web.bind.annotation.ControllerAdvice
         logger.error(e.message)
-        return ErrorDetail(e.httpStatus, e.message!!).toResponseEntity()
+        return translatedError(e.message!!, e.httpStatus, request)
     }
 
-    @ExceptionHandler(ConstraintViolationException::class)
-    protected fun handleConstraintViolationException(e: ConstraintViolationException, request: HttpServletRequest): ResponseEntity<Any> {
+    @ExceptionHandler(PSQLException::class)
+    protected fun handlePSQLException(e: PSQLException, request: WebRequest): ResponseEntity<Any> {
         logger.error(e.message)
-        return ErrorDetail(HttpStatus.BAD_REQUEST, e.message!!).toResponseEntity()
+        return unexpectedError(e, HttpStatus.INTERNAL_SERVER_ERROR, request)
     }
 
-    @ExceptionHandler(PSQLException::class)
-    protected fun handlePSQLException(e: PSQLException, request: HttpServletRequest): ResponseEntity<Any> {
-        logger.error(e.message)
-        return ErrorDetail(HttpStatus.INTERNAL_SERVER_ERROR, "An unexpected error occurred.".appendErrorCodeInstructions())
-                .toResponseEntity()
+    private fun getBundle (request: WebRequest): ResourceBundle {
+        val language = request.getHeader("Accept-Language") ?: "en"
+        return ResourceBundle.getBundle("ErrorMessageBundle", Locale(language))
     }
 
-    private fun String.appendErrorCodeInstructions(): String {
-        val code = errorCodeGenerator.newCode()
-        val fullStopIfNeeded =  if (this.endsWith('.')) ""  else "."
-        return "$this$fullStopIfNeeded If you see this message while you are using ${appProperties.applicationTitle} at" +
-                " a workshop, please contact your workshop technical support and show them this code: $code. " +
-                "Otherwise please contact support at ${appProperties.supportEmail} and quote this code: $code"
+    private fun translatedError(key: String, status: HttpStatus, request: WebRequest): ResponseEntity<Any> {
+        val resourceBundle = getBundle(request)
+        val message = if (resourceBundle.containsKey(key)) {
+            resourceBundle.getString(key)
+        } else {
+            key
+        }
+        return ErrorDetail(status, message).toResponseEntity()
+    }
+
+    private fun unexpectedError(e: Exception, status: HttpStatus, request: WebRequest): ResponseEntity<Any> {
+        val resourceBundle = getBundle(request)
+        var message = resourceBundle.getString("unexpectedError")
+        val formatter = MessageFormat(message, resourceBundle.locale)
+        val messageArguments = arrayOf(
+                appProperties.applicationTitle,
+                errorCodeGenerator.newCode(),
+                appProperties.supportEmail
+        )
+        message = formatter.format(messageArguments)
+
+        val originalMessage = if (!e.message.isNullOrEmpty()) {
+            listOf(e.message!!)
+        } else {
+            null
+        }
+        return ErrorDetail(status, message, originalMessage)
```

Comment:
Using the `trace` field here to return the original error message, e.g. "no handler found" etc. It's not actually the trace of course...